File: twoSum.js
// https://leetcode.com/problems/two-sum/

const twoSum = function(nums, target) {
	const storage = {};
	for (let i = 0; i < nums.length; i++) {
		let opp = target - nums[i];
		if (storage[opp] !== undefined) {
			return [storage[opp], i];
		}
		storage[nums[i]] = i;
	}
	console.log(storage);
}

console.assert(JSON.stringify(twoSum([1, 2, 7, 11, 15], 9)) === JSON.stringify([1, 2]), "works for sorted array", twoSum([1, 2, 7, 11, 15], 9));
console.assert(JSON.stringify(twoSum([7, 3, 2, 1, 15, 11], 9)) === JSON.stringify([0, 2]), "works for unsortedArray", twoSum([7, 3, 2, 1, 15, 11], 9));

File: twoSum.py
# https://leetcode.com/problems/two-sum/

def twoSumExtraSpace(nums, target):
	map = {}

	for index, value in enumerate(nums):
		if target - value in map:
			return (map[target - value], index)
		else:
			map[value] = index


def twoSum(nums, target):
	low, high = 0, len(nums) - 1

	while low < high:
		current = nums[low] + nums[high]

		if current == target:
			return (low, high)

		if current < target:
			low += 1
		else: 
			high -= 1
	
	return None


print(twoSum([2, 7, 11, 20], 9), (0, 1))
print(twoSum([2, 7, 11, 20], 31), (2, 3))
print(twoSum([2, 7, 11, 20], 22), (0, 3))
print(twoSum([2, 7, 11, 20], 18), (1, 2))
print(twoSumExtraSpace([2, 7, 11, 20], 9), (0, 1))
print(twoSumExtraSpace([2, 7, 11, 20], 31), (2, 3))
print(twoSumExtraSpace([2, 7, 11, 20], 22), (0, 3))
print(twoSumExtraSpace([2, 7, 11, 20], 18), (1, 2))
